File: server/llm/schemas.py
import uuid
from datetime import datetime
from typing import Literal

from common.schemas import CreatedResponse, OKResponse
from pydantic import BaseModel, Field, field_validator

AssistantMessageRole = Literal["assistant"]
MessageRoles = Literal["user"] | AssistantMessageRole


class LLMMessage(BaseModel):
    role: MessageRoles
    content: str


class ChatRoomMessage(LLMMessage):
    id: uuid.UUID
    llm_provider: str
    llm_key: str
    date: datetime

    @property
    def agent_name(self) -> str:
        return f"{self.llm_provider}:{self.llm_key}"

    @property
    def as_llm_message(self) -> LLMMessage:
        return LLMMessage(role=self.role, content=self.content)


class CreateChatMessagePayload(BaseModel):
    message: str = Field(..., min_length=1)

    @field_validator("message", mode="before")
    @classmethod
    def strip_whitespaces(cls, v: str) -> str:
        return v.strip()


class CreateChatRoomPayload(BaseModel):
    question: ChatRoomMessage
    answer: ChatRoomMessage


class CreateChatMessageResponse(CreatedResponse, ChatRoomMessage):
    room_id: uuid.UUID
    title: str
    updated_at: datetime


class ListChatMessagesResponse(OKResponse):
    data: list[ChatRoomMessage]

File: frontend/llm/schemas.ts
import * as z from 'zod';

import { DefaultResponseSchema } from '../common/api/client';

export type LLMMessage = z.infer<typeof LLMMessageSchema>;

export type GetMessagesResponse = z.infer<typeof GetMessagesResponseSchema>;

export type CreateMessageResponse = z.infer<typeof CreateMessageResponseSchema>;

const LLMMessageSchemaShape = {
  id: z.uuidv4(),
  llm_provider: z.string(),
  llm_key: z.string(),
  date: z.iso.datetime(),
  role: z.literal(['user', 'assistant']),
  content: z.string(),
};

export const LLMMessageSchema = z.object(LLMMessageSchemaShape);

export const GetMessagesResponseSchema = DefaultResponseSchema.extend({ data: z.array(LLMMessageSchema) });

export const CreateMessageResponseSchema = DefaultResponseSchema.extend(LLMMessageSchemaShape).extend({
  room_id: z.uuidv4(),
  title: z.string(),
  updated_at: z.iso.datetime(),
});
